### ISSUE TITLE
Unable to convert pipeline

### ISSUE THREAD

**Issue by Jimus17** [samsmithnz/AzurePipelinesToGitHubActionsConverter#128]
Pasting in azure devops pipeline YAML to https://pipelinestoactions.azurewebsites.net/ and clicking "process YAML" fails to produce any output or error message

1. Go to https://pipelinestoactions.azurewebsites.net/ 
2. paste pipeline YAML in to  'Azure DevOps Pipeline YAML'
3. Click 'process YAML'
4. pipeline YAML is reformatted but no output appears in the 'GitHub Actions YAML' box and no error message 

**Expected behavior**
GitHub Actions YAML box populated with something

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Browser: Chrome 84.0.4147.89, Firefox 79.0

**Screenshots**
![pipelineConvert](https://user-images.githubusercontent.com/8372842/89176770-100d6700-d582-11ea-8166-3355001cb1c3.PNG)

Example YAML
```
trigger:
- master

pool: 'Pipeline-Demo-Windows'


variables:
  solution: '**/*.sln'
  buildPlatform: 'Any CPU'
  buildConfiguration: 'Release'

stages:

- stage: Build
  jobs: 
    - job: BuildSpark
      pool:
        name: 'Pipeline-Demo-Windows'
        demands:
        - Agent.OS -equals Windows_NT
      steps:
      - task: NuGetToolInstaller@1

      - task: NuGetCommand@2
        inputs:
          restoreSolution: '$(solution)'

      - task: VSBuild@1
        inputs:
          solution: '$(solution)'
          msbuildArgs: '/p:DeployOnBuild=true /p:WebPublishMethod=Package /p:PackageAsSingleFile=true /p:SkipInvalidConfigurations=true /p:PackageLocation="$(build.artifactStagingDirectory)"'
          platform: '$(buildPlatform)'
          configuration: '$(buildConfiguration)'

      - task: PublishPipelineArtifact@1
        inputs:
          targetPath: '$(build.artifactStagingDirectory)'
          artifact: 'WebDeploy'
          publishLocation: 'pipeline'

- stage: Deploy
  jobs:
    - deployment: 
      variables:
        - name: Art
          value: "Server=.;Database=Art;Trusted_Connection=True;"
        
      environment: 
        name: windows-server
        resourceType: VirtualMachine
        tags: web
      strategy:
        runOnce:
          deploy:
            steps:
              - task: DownloadPipelineArtifact@2
                inputs:
                  buildType: 'current'
                  artifactName: 'WebDeploy'
                  targetPath: '$(Pipeline.Workspace)'


              - task: CmdLine@2
                inputs:
                  script: |
                    echo Write your commands here
                    
                    DIR
                  workingDirectory: '$(Pipeline.Workspace)'
                  
              - task: IISWebAppManagementOnMachineGroup@0
                inputs:
                  IISDeploymentType: 'IISWebsite'
                  ActionIISWebsite: 'CreateOrUpdateWebsite'
                  WebsiteName: 'Spark'
                  WebsitePhysicalPath: '%SystemDrive%\inetpub\wwwroot'
                  WebsitePhysicalPathAuth: 'WebsiteUserPassThrough'
                  AddBinding: true
                  CreateOrUpdateAppPoolForWebsite: true
                  ConfigureAuthenticationForWebsite: true
                  AppPoolNameForWebsite: 'Spark'
                  DotNetVersionForWebsite: 'v4.0'
                  PipeLineModeForWebsite: 'Integrated'
                  AppPoolIdentityForWebsite: 'ApplicationPoolIdentity'
                  AnonymousAuthenticationForWebsite: true
                  WindowsAuthenticationForWebsite: false
                  protocol: 'http' 
                  iPAddress: 'All Unassigned'
                  port: '80'
                  
              - task: IISWebAppDeploymentOnMachineGroup@0
                inputs:
                  WebSiteName: 'Spark'
                  Package: '$(Pipeline.Workspace)\Art.Web.zip'
                  XmlVariableSubstitution: true
```
          
    `

**Comment by samsmithnz** [comment]
Thanks @Jimus17. You've hit a sweet spot of a couple issues here. I've fixed a regression we added in the last release that was blocking errors from being returned to the client - at least now you get some feedback (although admittedly, the error YamlDotNet is a little garbled sometimes).

I've identified a few issues, but I need some time to resolve and harden the unit tests. The root cause is how Azure Pipelines YAML has variations for a few different properties, for example, pool can be either:

```
pool:
  name: string  # name of the pool to run this job in
  demands: string | [ string ]  # see the following "Demands" topic
  vmImage: string # name of the VM image you want to use; valid only in the Microsoft-hosted pool
```
OR:

```
pool: string # name of the private pool to run this job in
```

Obviously these are completely different data structures, with adds a lot of complexity to our code. (Currently I support the top version of pool, with a single string for demands, not a list array)


I hope to resolve the rest of this over the next week, and unblock you in the next day or two (I believe if we tweak some of your syntax we can workaround it before I code up a solution that can handle your variations).

**Comment by Jimus17** [comment]
Thanks @samsmithnz Having the error message makes it easier to see what's going on and to work out if massaging some of the formatting on the Azure DevOps side will make things work. This was a fairly simple pipeline, I have some more complex examples if that will help you debug / identify issues?

**Comment by samsmithnz** [comment]
If you are happy to share other pipelines, I would be happy to take them and add them to my list. At the very least I have a folder of YAML files I run over as part of the regression. 

Thanks for your patience, and submitting this issue. Really helpful with prioritization

**Comment by samsmithnz** [comment]
The final two other issues to address this have been identified. 
1. The environment syntax. I'm addressing that in PR #135
1. The variables syntax in the deployment job is more difficult to resolve. I've created a separate issue #134 to track that. Until then, the following syntax needs to be refactored, especially in a job: 
```
      variables:
      - name: MyVar
        value: MyValue
```

To this. This won't work if a template is involved too...
```
      variables:
        MyVar: MyValue
```

**Comment by samsmithnz** [comment]
**@Jimus17, you've probably moved on, but this is largely fixed, except for two small issues. I'm going to close this issue, as those items are tracked else where**. Thank you for submitting your pipeline, it's made the conversion a lot better, and highlighted some edge cases we handle better now.

1. Variables. You used an rarely seen, long form format, that you should only need if you are using variable groups. I recommend you update this:
    ```
          variables:
          - name: MyVar
            value: MyValue
    ```
    To this: 
    ```
          variables:
            MyVar: MyValue 
    ```
    I've created a new issue #134 to track this - I had only accounted for this at the top, pipeline level, not in a job, as I felt it was only needed as an exception when using variable groups. I need to find a way to identify this better.
2. A few tasks I don't have conversions for yet. 
   - IISWebAppDeploymentOnMachineGroup@0 and IISWebAppManagementOnMachineGroup@0: Created Issue #136 for this. I'm 90% focused on Azure these days, but have used this in the past and can see how it would be useful
   - DownloadPipelineArtifact@2: is likely very similar to the the DownloadBuildArtifacts@0 task that we do support. Created issue #138 for this.

Here is your pipeline with the variable changes recommended in #1 above: 

```
trigger:
- master

pool: 
  name: 'Pipeline-Demo-Windows'


variables:
  solution: '**/*.sln'
  buildPlatform: 'Any CPU'
  buildConfiguration: 'Release'

stages:

- stage: Build
  jobs: 
    - job: BuildSpark
      pool:
        name: 'Pipeline-Demo-Windows'
        demands:
        - Agent.OS -equals Windows_NT
      steps:
      - task: NuGetToolInstaller@1

      - task: NuGetCommand@2
        inputs:
          restoreSolution: '$(solution)'

      - task: VSBuild@1
        inputs:
          solution: '$(solution)'
          msbuildArgs: '/p:DeployOnBuild=true /p:WebPublishMethod=Package /p:PackageAsSingleFile=true /p:SkipInvalidConfigurations=true /p:PackageLocation="$(build.artifactStagingDirectory)"'
          platform: '$(buildPlatform)'
          configuration: '$(buildConfiguration)'

      - task: PublishPipelineArtifact@1
        inputs:
          targetPath: '$(build.artifactStagingDirectory)'
          artifact: 'WebDeploy'
          publishLocation: 'pipeline'

- stage: Deploy
  jobs:
    - deployment: 
      variables:
        Art: "Server=.;Database=Art;Trusted_Connection=True;"
        
      environment: 
        name: windows-server
        resourceType: VirtualMachine
        tags: 
          - web
      strategy:
        runOnce:
          deploy:
            steps:
              - task: DownloadPipelineArtifact@2
                inputs:
                  buildType: 'current'
                  artifactName: 'WebDeploy'
                  targetPath: '$(Pipeline.Workspace)'


              - task: CmdLine@2
                inputs:
                  script: |
                    echo Write your commands here
                    
                    DIR
                  workingDirectory: '$(Pipeline.Workspace)'
                  
              - task: IISWebAppManagementOnMachineGroup@0
                inputs:
                  IISDeploymentType: 'IISWebsite'
                  ActionIISWebsite: 'CreateOrUpdateWebsite'
                  WebsiteName: 'Spark'
                  WebsitePhysicalPath: '%SystemDrive%\inetpub\wwwroot'
                  WebsitePhysicalPathAuth: 'WebsiteUserPassThrough'
                  AddBinding: true
                  CreateOrUpdateAppPoolForWebsite: true
                  ConfigureAuthenticationForWebsite: true
                  AppPoolNameForWebsite: 'Spark'
                  DotNetVersionForWebsite: 'v4.0'
                  PipeLineModeForWebsite: 'Integrated'
                  AppPoolIdentityForWebsite: 'ApplicationPoolIdentity'
                  AnonymousAuthenticationForWebsite: true
                  WindowsAuthenticationForWebsite: false
                  protocol: 'http' 
                  iPAddress: 'All Unassigned'
                  port: '80'
                  
              - task: IISWebAppDeploymentOnMachineGroup@0
                inputs:
                  WebSiteName: 'Spark'
                  Package: '$(Pipeline.Workspace)\Art.Web.zip'
                  XmlVariableSubstitution: true
```
